db/db_impl/db_impl.cc

```diff
@@ -2220,6 +2218,12 @@ Status DBImpl::GetImpl(const ReadOptions& read_options, const Slice& key,
     RecordTick(stats_, NUMBER_KEYS_READ);
     size_t size = 0;
     if (s.ok()) {
+      const size_t merge_threshold = read_options.merge_operand_count_threshold;
+      if (merge_threshold > 0 &&
+          merge_context.GetNumOperands() >= merge_threshold) {
+        s = Status::OkMergeOperandThresholdReached();
```

Comment:
No strong opinion on this, but I'm wondering if we want to name this `OkMergeOperandThresholdExceeded` and check `merge_context.GetNumOperands() > merge_threshold` instead.

This sort of aligns with the context like rate limiting. If the rate reaches the limit exactly, it's still OK, but the client cannot exceed any single byte after that.

Reply:
Yeah, I thought about that too - the reason I went with this is that I wanted to be in sync with `max_successive_merges` where 0 means disabled. (Using strictly-greater semantics would imply using a signed type for the option and negatives for disabled.)

Reply:
I see that there's an option, `get_impl_options.get_merge_operands_options->expected_max_number_of_operands`. This isn't available in `MultiGet`.

Unlike `merge_operand_count_threshold`, it appears to be a hard limit. (returns Status::Incomplete(Status::SubCode::KMergeOperandsInsufficientCapacity) but I think it serves the same purpose as it also sets the status when num_of_operands is greater than the option value.

I am not sure if this option is being used much outside of the tests. I am wondering if we want to clean up `GetMergeOperandsOptions` completely and replace it with your new option here.

Reply:
> Yeah, I thought about that too - the reason I went with this is that I wanted to be in sync with `max_successive_merges` where 0 means disabled. (Using strictly-greater semantics would imply using a signed type for the option and negatives for disabled.)

... or I suppose we can go for a more modern C++ solution and use `std::optional`. Actually, let me do that

Reply:
> I see that there's an option, `get_impl_options.get_merge_operands_options->expected_max_number_of_operands`. This isn't available in `MultiGet`.
> 
> Unlike `merge_operand_count_threshold`, it appears to be a hard limit. (returns Status::Incomplete(Status::SubCode::KMergeOperandsInsufficientCapacity) but I think it serves the same purpose as it also sets the status when num_of_operands is greater than the option value.
> 
> I am not sure if this option is being used much outside of the tests. I am wondering if we want to clean up `GetMergeOperandsOptions` completely and replace it with your new option here.

Thanks for bringing this up! I wasn't aware of this option. Will take a closer look to see if there's room for some unification here but when it comes to the `Status` subcode and `GetMergeOperands` specifically, we can't get rid of them for now (at least until the next major release) since they're in the public API and that would be a breaking change. (`GetMergeOperandsOptions` is an argument of `DB::GetMergeOperands`.)

Reply:
>specifically, we can't get rid of them for now (at least until the next major release) since they're in the public API and that would be a breaking change. (GetMergeOperandsOptions is an argument of DB::GetMergeOperands.)

Yeap. If we are going to deprecate the old option by replacing with the new one and change the behavior to be a soft-limit, it's inevitably a breaking change + change of behavior that could be very disruptive to users.

I guess since the hard limit is only applicable to `DB::GetMergeOperands`, users won't be confused by two different options with the similar meaning in most cases, and we don't have to worry about combining them, but maybe it's worthwhile to add quick comment in `GetMergeOperandsOptions` that it has a different behavior than this new `merge_operand_count_threshold` in read_option.